File: 35.搜素插入位置/searchInsert.go
package problem0035

// 排序数组 无重复, 有目标值则返回索引, 无目标值返回插入索引
// 普通做法, O(n)时间复杂度
func searchInsert(nums []int, target int) int {
	for i, v := range nums {
		if target <= v {
			return i
		}
	}
	return len(nums)
}

// 二分查找的话, 可以降低到O(logn)的时间复杂度
func searchInsertA(nums []int, target int) int {
	left, right := 0, len(nums)

	for left < right {
		mid := (left + right) >> 1
		v := nums[mid]

		if v >= target {
			right = mid
		} else {
			left = mid + 1
		}
	}
	return left
}

File: 35.搜素插入位置/searchInsert.js
function searchInsert(nums, target) {
	// 特殊情况处理
	const n = nums.length
	if (target > nums[n - 1]) {
		return n
	}
	// 在[0, n - 1]中一定有解
	let left = 0
	let right = n - 1
	while (left < right) {
		const mid = left + Math.floor((right - left) / 2)
		// 需要寻找 >= target 的第一个索引
		if (nums[mid] < target) {
			left = mid + 1
		} else {
			right = mid
		}
	}
	return left
}